crates/red_knot_test/src/matcher.rs

```diff
@@ -0,0 +1,731 @@
+                        // diagnostics as all unmatched, and advance the diagnostics iterator.
+                        failures.insert(diagnostics.line, unmatched(diagnostics));
+                        current_diagnostics = line_diagnostics.next();
+                    }
+                }
+            }
+            (Some(assertions), None) => {
+                // We've exhausted diagnostics but still have assertions; report these assertions
+                // as unmatched and advance the assertions iterator.
+                failures.insert(assertions.line, unmatched(assertions));
+                current_assertions = line_assertions.next();
+            }
+            (None, Some(diagnostics)) => {
+                // We've exhausted assertions but still have diagnostics; report these
+                // diagnostics as unmatched and advance the diagnostics iterator.
+                failures.insert(diagnostics.line, unmatched(diagnostics));
+                current_diagnostics = line_diagnostics.next();
+            }
+            // When we've exhausted both diagnostics and assertions, break.
+            (None, None) => break,
+        }
+    }
+
+    if failures.is_empty() {
+        Ok(())
+    } else {
+        Err(failures)
+    }
+}
+
+pub(super) trait Diagnostic: Ranged {
+    fn rule(&self) -> &str;
+
+    fn message(&self) -> &str;
+}
+
+impl Diagnostic for Arc<TypeCheckDiagnostic> {
+    fn rule(&self) -> &str {
+        self.as_ref().rule()
+    }
+
+    fn message(&self) -> &str {
+        self.as_ref().message()
+    }
+}
+
+trait Unmatched {
+    fn unmatched(&self) -> String;
+}
+
+impl<T> Unmatched for T
+where
+    T: Diagnostic,
+{
+    fn unmatched(&self) -> String {
+        format!(
+            r#"unexpected error: [{}] "{}""#,
+            self.rule(),
+            self.message()
+        )
+    }
+}
+
+impl Unmatched for Assertion<'_> {
+    fn unmatched(&self) -> String {
+        format!("unmatched assertion: {self}")
+    }
+}
+
+fn unmatched<'a, T: Unmatched + 'a>(unmatched: impl Into<&'a [T]>) -> Vec<String> {
+    unmatched.into().iter().map(Unmatched::unmatched).collect()
+}
+
+struct Matcher {
+    line_index: LineIndex,
+    source: SourceText,
+}
+
+impl Matcher {
+    fn from_file(db: &TestDb, file: File) -> Self {
+        Self {
+            line_index: line_index(db, file),
+            source: source_text(db, file),
+        }
+    }
+
+    /// Check a slice of [`Diagnostic`]s against a slice of [`Assertion`]s.
+    ///
+    /// Return vector of [`Unmatched`] for any unmatched diagnostics or assertions.
+    fn match_line<'a, 'b, T: Diagnostic + 'a>(
+        &self,
+        diagnostics: impl Into<&'a [T]>,
+        assertions: impl Into<&'a [Assertion<'b>]>,
+    ) -> Result<(), Vec<String>>
+    where
+        'b: 'a,
+    {
+        let mut failures = vec![];
+        let mut unmatched = diagnostics.into().iter().collect::<Vec<_>>();
```

Comment:
It's not clear to me how this helps avoid the need for the `matched` hashset in `matches`. I would assume you still can't pop from an `FxHashSet` while iterating over it?

Reply:
I would have do take a closer look again. You can use `retain` which iterators for you

Reply:
I already use `retain` with a vec. The `O(assertions)` hashsets are to track the indices to remove, which I collect while iterating and then later need to use within my call to `retain`.

Reply:
I took another look at `Matcher::matches` and was able to rework it slightly to avoid the per-assertion hashset allocation there. I still have `unmatched` as a Vec, though.